compiler/luci/profile/src/CircleNodeOrigin.cpp

```diff
@@ -0,0 +1,161 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Profile/CircleNodeOrigin.h"
+
+#include <loco.h>
+
+#include <cassert>
+#include <vector>
+
+namespace
+{
+
+/**
+ * @brief Set annotation for recording origin information
+ * @note  Once CircleNodeOrigin is annotated, it should not be changed.
+ *        If CircleNodeOrigin is needed to be changed, create new CircleNodeOrigin.
+ */
+class CircleNodeOriginAnnotation final : public loco::NodeAnnotation
+{
+public:
+  CircleNodeOriginAnnotation() = delete;
+
+  CircleNodeOriginAnnotation(const std::shared_ptr<luci::CircleNodeOrigin> origin) : _origin(origin)
+  {
+    // Do nothing
+  }
+
+public:
+  const std::shared_ptr<luci::CircleNodeOrigin> origin(void) const { return _origin; }
+  // No setter
+
+private:
+  const std::shared_ptr<luci::CircleNodeOrigin> _origin;
+};
+
+} // namespace
+
+namespace
+{
+
+class SingleOrigin final : public luci::CircleNodeOrigin
+{
+public:
+  SingleOrigin() = delete;
+
+  SingleOrigin(uint32_t id, const std::string &name)
+  {
+    _source.id(id);
+    _source.name(name);
+  }
+
+public:
+  std::set<const Source *> sources(void) const final
+  {
+    std::set<const Source *> res;
+    res.emplace(&_source);
+    return res;
+  }
+
+private:
+  Source _source;
+};
+
+class CompositeOrigin final : public luci::CircleNodeOrigin
+{
+public:
+  CompositeOrigin() = delete;
+
+  template <typename T> CompositeOrigin(T origins)
+  {
+    if (origins.size() == 0)
+      throw std::invalid_argument("No origins provided");
+
+    for (auto &origin : origins)
+    {
+      if (origin != nullptr)
+        _origins.emplace_back(origin);
+    }
+  }
+
+public:
+  std::set<const Source *> sources(void) const final
+  {
+    std::set<const Source *> res;
+
+    for (auto &origin : _origins)
+    {
+      for (auto source : origin->sources())
+      {
+        res.emplace(source);
+      }
+    }
+
+    return res;
+  }
+
+private:
+  std::vector<std::shared_ptr<CircleNodeOrigin>> _origins;
+};
+
+} // namespace
+
+namespace luci
+{
+
+std::shared_ptr<CircleNodeOrigin> single_origin(uint32_t id, const std::string name)
+{
+  return std::make_shared<SingleOrigin>(id, name);
+}
+
+std::shared_ptr<CircleNodeOrigin>
+composite_origin(std::initializer_list<std::shared_ptr<CircleNodeOrigin>> origins)
+{
+  return std::make_shared<CompositeOrigin>(origins);
+}
+
+std::shared_ptr<CircleNodeOrigin>
+composite_origin(std::vector<std::shared_ptr<CircleNodeOrigin>> origins)
```

Comment:
Q) there is `composite_origin` with `initializer_list` . why is also `vector` needed?

Reply:
Q) why did you not use `&origins` ?

Reply:
> Q) there is composite_origin with initializer_list . why is also vector needed?

When the number of origins is fixed, we can use `initializer_list` and it makes code simple and easy.
(For example, when we propagate origins in `luci/Pass`)

However, when the number is not fixed, we cannot use `initializer_list` :(
(For example, when we import circle file and annotate in `luci/import`)

By that reason, both of them are introduced.

Reply:
> However, when the number is not fixed, we cannot use initializer_list :(

Can you show me a draft code for this? It's a bit hard to understand the case.

Reply:
Following codes use `composite_origin(std::vector<std::shared_ptr<CircleNodeOrigin>> origins)` at #6101.

https://github.com/Samsung/ONE/blob/9f6a31ee0a1d5d024b82cd37acda3bf741ded779/compiler/luci/import/src/ImportMetadata.cpp#L144-L154

As we do not know the size of `source_ids`, we cannot use `initializer_list` for that case.
However, in case of `luci/Pass`, we can know the exact number of origins.
Please refer to following draft codes in #6217.

https://github.com/Samsung/ONE/blob/fc3b7f724b9aa69c0a78adafa2ac431d56f4b0fb/compiler/luci/pass/src/FuseBatchNormWithConvPass.cpp#L204-L205